server/src/main/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkers.java

```diff
@@ -0,0 +1,179 @@
+            this.context = context;
+        }
+
+        public SnapshotShardContext context() {
+            return context;
+        }
+
+        public void run() {
+            shardSnapshotter.accept(context);
+        }
+
+        public short priority() {
+            return 1;
+        }
+
+        @Override
+        public int compareTo(ShardSnapshotTask other) {
+            int res = context.snapshotId().compareTo(other.context.snapshotId());
+            if (res != 0) {
+                return res;
+            }
+            return Integer.compare(priority(), other.priority());
+        }
+    }
+
+    class FileSnapshotTask extends ShardSnapshotTask {
+        private final FileInfo fileInfo;
+        private final ActionListener<Void> fileUploadListener;
+
+        FileSnapshotTask(SnapshotShardContext context, FileInfo fileInfo, ActionListener<Void> fileUploadListener) {
+            super(context);
+            this.fileInfo = fileInfo;
+            this.fileUploadListener = fileUploadListener;
+        }
+
+        @Override
+        public void run() {
+            ActionRunnable.run(fileUploadListener, () -> fileSnapshotter.accept(context, fileInfo)).run();
+        }
+
+        @Override
+        public short priority() {
+            return 2;
+        }
+    }
+
+    public ShardSnapshotWorkers(
+        final int maxWorkers,
+        final Executor executor,
+        final Consumer<SnapshotShardContext> shardSnapshotter,
+        final CheckedBiConsumer<SnapshotShardContext, FileInfo, IOException> fileSnapshotter
+    ) {
+        assert maxWorkers > 0;
+
+        logger.info("starting shard snapshot worker pool of max size {}", maxWorkers);
+        this.maxWorkers = maxWorkers;
+        this.executor = executor;
+        this.shardSnapshotter = shardSnapshotter;
+        this.fileSnapshotter = fileSnapshotter;
+    }
+
+    public void enqueueShardSnapshot(final SnapshotShardContext context) {
+        logger.trace("enqueuing shard snapshot task [snapshotID={}, indexID={}]", context.snapshotId(), context.indexId());
+        synchronized (mutex) {
+            snapshotTasks.add(new ShardSnapshotTask(context));
+            ensureEnoughWorkers();
+        }
+    }
+
+    public void enqueueFileUpload(
+        final SnapshotShardContext context,
+        final FileInfo fileInfo,
+        final ActionListener<Void> fileUploadListener
+    ) {
+        logger.trace(
+            "enqueuing shard snapshot file upload task [snapshotID={}, indexID={}, file={}]",
+            context.snapshotId(),
+            context.indexId(),
+            fileInfo.name()
+        );
+        synchronized (mutex) {
+            snapshotTasks.add(new FileSnapshotTask(context, fileInfo, fileUploadListener));
+            ensureEnoughWorkers();
+        }
+    }
+
+    private void ensureEnoughWorkers() {
+        assert Thread.holdsLock(mutex);
+
+        int workersToCreate = Math.min(maxWorkers - workerCount, snapshotTasks.size());
```

Comment:
I think we would max add one worker here ever? Also, using `snapshotTasks.size()` seems wrong, since the task is polled by a worker, but that worker is not free to do other work until it completes it's task (and thus we may still want a new worker even if there is only 1 task in the queue).

Also, as pointed out by Armin, we poll outside the mutex anyway.

I wonder if we could instead just start a worker every time as long as we have budget for it?

Reply:
> I wonder if we could instead just start a worker every time as long as we have budget for it?

++ there isn't much value in being tricky here, just starting up to pool-size workers seems like a good idea and removes the rather hard to reason about `size()` call.

Reply:
Hmm. I don't really see that happening as of https://github.com/pxsalehi/elasticsearch/blob/5397f5207da55b41b33afb10046d3079833829a1/server/src/main/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkers.java. Maybe I'm missing something. In your example, since `maxWorker` is 2 and one is busy, when the second task gets queued we'd call `ensureEnoughWorkers` and we'd end up with `min(2-1, 1)` workers, so a new one will be started. (I think [this test](https://github.com/pxsalehi/elasticsearch/blob/3649729263f79e3d80d191406008e4a6e22df688/server/src/test/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkersTests.java#L157-L181) kind of covers a similar case).